// @cratecode/client/dist/proto_video/change_toolbox.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ChangeToolbox = exports.ToolboxType = void 0;
const runtime_1 = require("@protobuf-ts/runtime");
const runtime_2 = require("@protobuf-ts/runtime");
const runtime_3 = require("@protobuf-ts/runtime");
const runtime_4 = require("@protobuf-ts/runtime");
const runtime_5 = require("@protobuf-ts/runtime");
/**
 * @generated from protobuf enum ToolboxType
 */
var ToolboxType;
(function (ToolboxType) {
    /**
     * @generated from protobuf enum value: Info = 0;
     */
    ToolboxType[ToolboxType["Info"] = 0] = "Info";
    /**
     * @generated from protobuf enum value: Files = 1;
     */
    ToolboxType[ToolboxType["Files"] = 1] = "Files";
})(ToolboxType = exports.ToolboxType || (exports.ToolboxType = {}));
// @generated message type with reflection information, may provide speed optimized methods
class ChangeToolbox$Type extends runtime_5.MessageType {
    constructor() {
        super("ChangeToolbox", [
            { no: 1, name: "type", kind: "enum", T: () => ["ToolboxType", ToolboxType] }
        ]);
    }
    create(value) {
        const message = { type: 0 };
        globalThis.Object.defineProperty(message, runtime_4.MESSAGE_TYPE, { enumerable: false, value: this });
        if (value !== undefined)
            (0, runtime_3.reflectionMergePartial)(this, message, value);
        return message;
    }
    internalBinaryRead(reader, length, options, target) {
        let message = target !== null && target !== void 0 ? target : this.create(), end = reader.pos + length;
        while (reader.pos < end) {
            let [fieldNo, wireType] = reader.tag();
            switch (fieldNo) {
                case /* ToolboxType type */ 1:
                    message.type = reader.int32();
                    break;
                default:
                    let u = options.readUnknownField;
                    if (u === "throw")
                        throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
                    let d = reader.skip(wireType);
                    if (u !== false)
                        (u === true ? runtime_2.UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
            }
        }
        return message;
    }
    internalBinaryWrite(message, writer, options) {
        /* ToolboxType type = 1; */
        if (message.type !== 0)
            writer.tag(1, runtime_1.WireType.Varint).int32(message.type);
        let u = options.writeUnknownFields;
        if (u !== false)
            (u == true ? runtime_2.UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
        return writer;
    }
}
/**
 * @generated MessageType for protobuf message ChangeToolbox
 */
exports.ChangeToolbox = new ChangeToolbox$Type();
