// @cratecode/client/dist/upload/unit.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.handleUnit = void 0;
const axios_1 = __importDefault(require("axios"));
const index_1 = require("./index");
/**
 * Handles a unit manifest and returns its actual ID.
 * @param state {State} - is the application's state.
 * @param id {string} - is the friendly name of the unit.
 * @param name {string} - is the display name of the unit.
 * @param lessons {Record<string, object>} - is the lessons map for the unit.
 */
function handleUnit(state, id, name, lessons) {
    var _a, _b;
    return __awaiter(this, void 0, void 0, function* () {
        // First, we need to figure out what the actual ID of our unit is.
        // If there isn't one, we'll just set it to null.
        const actualID = yield axios_1.default
            .get("https://cratecode.com/internal/api/id/" + id, {
            headers: {
                authorization: state.key,
            },
        })
            .then((res) => res.data.id)
            .catch((e) => {
            var _a;
            // If none was found, just use null.
            if (((_a = e.response) === null || _a === void 0 ? void 0 : _a.status) === 404)
                return null;
            throw e;
        });
        yield (0, index_1.delay)(state);
        // Next, we need to go through the lessons and map from friendly names to ids.
        const map = {};
        for (const lessonID in lessons) {
            // Map the key.
            const newKey = yield mapID(lessonID, state);
            // Map next and previous.
            const next = (_a = lessons[lessonID].next) !== null && _a !== void 0 ? _a : [];
            const previous = (_b = lessons[lessonID].previous) !== null && _b !== void 0 ? _b : [];
            if (typeof next !== "object" || !Array.isArray(next))
                throw new Error("next must be a string array!");
            if (typeof previous !== "object" || !Array.isArray(previous))
                throw new Error("previous must be a string array!");
            const newNext = [];
            const newPrevious = [];
            for (const item of next) {
                newNext.push(yield mapID(item, state));
            }
            for (const item of previous) {
                newPrevious.push(yield mapID(item, state));
            }
            map[newKey] = {
                next: newNext,
                previous: newPrevious,
                requireAll: Boolean(lessons[lessonID].requireAll),
            };
        }
        // Create or update the unit.
        const unitID = yield axios_1.default.put("https://cratecode.com/internal/api/unit", {
            id: actualID,
            friendlyName: id,
            name,
            data: map,
        }, {
            headers: {
                authorization: state.key,
            },
        }).then((res) => res.data.id);
        yield (0, index_1.delay)(state);
        // And finally, return the unit ID.
        return unitID;
    });
}
exports.handleUnit = handleUnit;
/**
 * Maps a friendly name or ID to an ID.
 * @param id {string} - is the ID or friendly name to lookup.
 * @param state {State} - is the application's state.
 */
function mapID(id, state) {
    return __awaiter(this, void 0, void 0, function* () {
        // IDs starting with a ":" are actual IDs, so we should just remove the ":".
        if (id.startsWith(":"))
            return id.substring(1);
        const newKey = state.idsMap[id] ||
            (yield axios_1.default
                .get("https://cratecode.com/internal/api/id/" + id, {
                headers: {
                    authorization: state.key,
                },
            })
                .then((res) => {
                state.idsMap[id] = res.data.id;
                return res.data.id;
            }));
        yield (0, index_1.delay)(state);
        return newKey;
    });
}
